### ISSUE TITLE
Feature-Request: scan processes via PID

### ISSUE THREAD

**Issue by Anthony-76** [VirusTotal/yara-x#199]
Hi,

Maybe I post my request in wrong Place. Please excuse me .

I use Yara and it's a nice product.

I beguin to use Yara-X but an important feature for me is (seems to be)  not present. It's scan process (with PID number).

This feature will be integrated in the new release or not?

Best regards
Anthony

**Comment by plusvic** [comment]
It's not likely to be implemented anytime soon. There are a lot of other things higher in the list. There's a brief comment about this in the documentation: https://virustotal.github.io/yara-x/docs/intro/yara-x-vs-yara/#no-process-scanning

**Comment by Anthony-76** [comment]
Ok,

Thank for your comment.

If you implement this feature, I'm ready to test it

**Comment by TheoTurletti-Kyos** [comment]
Also really interested

**Comment by JonathanAnbary** [comment]
Gonna start working on this, let me know if anybody else has done some work previously.

**Comment by ruppde** [comment]
> Gonna start working on this, let me know if anybody else has done some work previously.

there's boreal, another yara scanner written in rust, which supports process scanning on Windows, Linux and macOS:
https://github.com/vthib/boreal
https://docs.rs/boreal/latest/boreal/scanner/struct.FragmentedScanMode.html

don't know if the licenses are compatible, to allow copying code or if it can only be used as inspiration? best way is probably to contact @vthib anyway.

**Comment by JonathanAnbary** [comment]
That is really cool @ruppde!
I dont think copying code will be necessary (or really possible) since the structure of the code is quite different and the parts related to process scanning are not overly complex (already had them close to implemented before seeing your comment).
The main difference for the purposes of an implementation in yara-x then in boreal is that boreal uses a Box<[u8]> for holding the content of a match while yara-x uses just an offset + length.
this is fine for file scanning where the entire file is held in memory at once (either via an mmap or a regular buffer/vector) so the match data can be retrieved without needing duplication.
but when scanning process memory it is not ideal to hold the entirety of the scanned process memory in a buffer/mmap so a Box<[u8]> might be necessary.
Although I say not ideal, its also possible to just create a large buffer (or more likely anonymous mmap) and write the memory of the scanned process into it.
There is also another issue of the semantics of module's + conditions that depend on offsets, which boreal address.
boreal offers 3 different scan modes:
1. match the behavior of yara (which is a bit unintuitive in my opinion and maybe not the most useful).
2. optimize for fast scanning by disabling file scanning modules and additional fetches of data (for things that depend on offsets),
3. ensure a single fetch of the data but still do module scanning like in yara.

In my opinion the fast mode makes the most sense (I have a hard time thinking of a use case for file scanning modules that work on a processes memory, and conditions that depend on offsets also dont make much sense to me in the context of of fragmented memory regions). 
I currently have an implementation of the fast mode (for windows and linux) with the caveat that both the no-scan optimization and retrieving the match data are still not implemented.

@plusvic it would be very helpful if you could weigh in on these questions:
1. should the Match object be made to hold a copy of the content in the case of fragmented regions scanning, or should the regions be kept in a single mmap/buffer (probably mmap since buffer fails above a certain size), or maybe the match data should be re-fetched if its accessed?
2. which of the scan modes listed by boreal do you think is best?

**Comment by plusvic** [comment]
Yes, I think that the match object should hold a copy of the content in the case of scanning fragmented memory regions, but that should be done only if necessary. The Rust `Cow` type may be useful to achieve this.

Regarding modules, I would disabled them for memory scanning. Making modules like `pe` or `elf` work in the context of a process scanning will be hard, so I think that process memory scanning should be only a about finding patterns in the process memory space.

**Comment by JonathanAnbary** [comment]
I'm not really sure I understood how the `Cow` type could be used in this case, but I've implemented the content fetch for the match object in a way that I think achieve the best of both worlds.
I track the ranges within each scanned data that matched and then I copy them from within the scanned data before it is overwritten with the next scanned data but only once per range (so matches that overlap do not require twice the memory).
This ensures that we retain the minimum amount of data in order to return the contents of matches, a potential optimization is saving the entirety of the current scanned data in case a significant percent of it is matched (saves on a copy, and potentially allocations if the matches are heavily fragmented).
I have not implemented the no-scan optimization, since modules and offset matches are disabled I do not think it is worth doing (will only effect the runtime of rules whose results should be known at compile time anyway), tell me if you think it should be done anyway.
(also have not implemented anything in the cli/bindings that would allow accessing scan_proc)
https://github.com/VirusTotal/yara-x/pull/384

**Comment by umbrageodotus** [comment]
Can I help in any way?

**Comment by JonathanAnbary** [comment]
Discussion and work have continued over there: https://github.com/VirusTotal/yara-x/pull/384
